src/types.rs

```diff
@@ -127,19 +127,54 @@ impl StackTrieNode {
 }
 
 #[derive(Eq, PartialEq, Hash, Deserialize, Serialize, Debug, Clone)]
-pub struct CompileId {
+pub struct DynamoId {
     pub frame_id: u32,
     pub frame_compile_id: u32,
     pub attempt: u32,
 }
 
+#[derive(Eq, PartialEq, Hash, Deserialize, Serialize, Debug, Clone)]
+#[serde(untagged)]
+pub enum CompileId {
+    // NOTE: serde(untagged) will match in the order
+    // this enum is defined 
+
+    // When compiled autograd calls torch.compile
+    CompiledAutogradInitiated {
+        compiled_autograd_id: u32,
+        #[serde(flatten)]
+        dynamo_id: Option<DynamoId>,
+    },
+    // When user calls torch.compile
+    UserInitiated(DynamoId)
```

Comment:
Why do you want to have an untagged union here? Why not just have a single struct with optional fields everywhere?

Reply:
(I am generally a big union stan but JSON is a lot more straightforward to deserialize and evolve if you don't do unions, since it has no built-in notion of union)

Reply:
Untagged to be backwards compatible with old logs. The single struct with optionals feels too permissive because it's not obvious from looking at the definition what set of CompileId you can expect